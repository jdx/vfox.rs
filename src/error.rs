use mlua::Error as MLuaError;
use thiserror::Error;
use xx::XXError;

#[derive(Error, Debug)]
#[non_exhaustive]
pub enum VFoxError {
    #[error(transparent)]
    LuaError(#[from] MLuaError),
    #[error("serde_json error")]
    SerdeJsonError(#[from] serde_json::Error),
    #[error(transparent)]
    XXError(#[from] XXError),
}

pub type Result<T> = std::result::Result<T, VFoxError>;
