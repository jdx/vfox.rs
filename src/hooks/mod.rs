mod available;
mod post_install;
mod pre_install;
