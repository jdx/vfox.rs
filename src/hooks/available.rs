use mlua::prelude::LuaError;
use mlua::{FromLua, Lua, Value};

use crate::error::Result;
use crate::Plugin;

impl Plugin {
    #[tokio::main(flavor = "current_thread")]
    pub async fn available(&self) -> Result<Vec<AvailableVersion>> {
        self.available_async().await
    }

    pub async fn available_async(&self) -> Result<Vec<AvailableVersion>> {
        let ctx = self.context(None)?;
        let available = self
            .eval_async(chunk! {
                require "hooks/available"
                return PLUGIN:Available($ctx)
            })
            .await?;

        Ok(available)
    }
}

#[derive(Debug)]
pub struct AvailableVersion {
    pub version: String,
    pub note: Option<String>,
    // pub addition: Option<Table<'lua>>,
}

impl<'lua> FromLua<'lua> for AvailableVersion {
    fn from_lua(value: Value<'lua>, _: &'lua Lua) -> std::result::Result<Self, LuaError> {
        match value {
            Value::Table(table) => {
                // TODO: try to default this to an empty table or something
                // let addition = table.get::<_, Option<Table>>("addition")?;
                Ok(AvailableVersion {
                    version: table.get::<_, String>("version")?,
                    note: table.get::<_, Option<String>>("note")?,
                    // addition,
                })
            }
            _ => panic!("Expected table"),
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::Plugin;

    #[test]
    fn dummy() {
        let versions = run("vfox-dummy");
        assert_debug_snapshot!(versions, @r###"
        [
            "1.0.0",
            "1.0.1",
        ]
        "###);
    }

    #[tokio::test]
    async fn dummy_async() {
        let versions = run_async("vfox-dummy").await;
        assert_debug_snapshot!(versions, @r###"
        [
            "1.0.0",
            "1.0.1",
        ]
        "###);
    }

    #[tokio::test]
    async fn nodejs_async() {
        let versions = run_async("vfox-nodejs").await;
        assert!(versions.contains(&"20.0.0".to_string()));
    }

    fn run(plugin: &str) -> Vec<String> {
        let p = Plugin::test(plugin);
        let r = p.available().unwrap();
        r.iter().map(|v| v.version.clone()).collect()
    }

    async fn run_async(plugin: &str) -> Vec<String> {
        let p = Plugin::test(plugin);
        let r = p.available_async().await.unwrap();
        r.iter().map(|v| v.version.clone()).collect()
    }
}
