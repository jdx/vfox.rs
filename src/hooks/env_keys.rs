use mlua::prelude::LuaError;
use mlua::{FromLua, IntoLua, Lua, Value};
use std::collections::BTreeMap;
use std::path::PathBuf;

use crate::error::Result;
use crate::sdk_info::SdkInfo;
use crate::Plugin;

#[derive(Debug)]
pub struct EnvKey {
    pub key: String,
    pub value: String,
}

#[derive(Debug)]
pub struct EnvKeysContext {
    pub version: String,
    pub path: PathBuf,
    pub sdk_info: BTreeMap<String, SdkInfo>,
}

impl Plugin {
    pub async fn env_keys(&self, ctx: EnvKeysContext) -> Result<Vec<EnvKey>> {
        let env_keys = self
            .eval_async(chunk! {
                require "hooks/env_keys"
                return PLUGIN:EnvKeys($ctx)
            })
            .await?;

        Ok(env_keys)
    }
}

impl<'lua> IntoLua<'lua> for EnvKeysContext {
    fn into_lua(self, lua: &'lua Lua) -> mlua::Result<Value<'lua>> {
        let table = lua.create_table()?;
        table.set("version", self.version)?;
        table.set("path", self.path.to_string_lossy().to_string())?;
        let sdk_info = lua.create_table()?;
        for (k, v) in self.sdk_info {
            sdk_info.set(k, v)?;
        }
        table.set("sdk_info", sdk_info)?;
        Ok(Value::Table(table))
    }
}

impl<'lua> FromLua<'lua> for EnvKey {
    fn from_lua(value: Value<'lua>, _: &'lua Lua) -> std::result::Result<Self, LuaError> {
        match value {
            Value::Table(table) => Ok(EnvKey {
                key: table.get::<_, String>("key")?,
                value: table.get::<_, String>("value")?,
            }),
            _ => panic!("Expected table"),
        }
    }
}
