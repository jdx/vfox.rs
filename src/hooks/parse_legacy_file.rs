use mlua::prelude::LuaError;
use mlua::{FromLua, IntoLua, Lua, MultiValue, Value};
use std::path::{Path, PathBuf};

use crate::error::Result;
use crate::plugin::AppData;
use crate::Plugin;

#[derive(Debug)]
pub struct LegacyFileContext {
    pub filepath: PathBuf,
}

#[derive(Debug)]
pub struct ParseLegacyFileResponse {
    pub version: Option<String>,
}

impl Plugin {
    pub async fn parse_legacy_file(&self, legacy_file: &Path) -> Result<ParseLegacyFileResponse> {
        let ctx = LegacyFileContext {
            filepath: legacy_file.to_path_buf(),
        };
        let legacy_file_response = self
            .eval_async(chunk! {
                require "hooks/available"
                require "hooks/parse_legacy_file"
                return PLUGIN:ParseLegacyFile($ctx)
            })
            .await?;

        Ok(legacy_file_response)
    }
}

impl<'lua> IntoLua<'lua> for LegacyFileContext {
    fn into_lua(self, lua: &'lua Lua) -> mlua::Result<Value<'lua>> {
        let table = lua.create_table()?;
        table.set("filepath", self.filepath.to_string_lossy().to_string())?;
        table.set(
            "getInstalledVersions",
            lua.create_async_function(move |lua, _input: MultiValue| async {
                let app_data = lua.app_data_ref::<AppData>().unwrap();
                Ok(Plugin::from_dir(app_data.plugin_dir.as_path())
                    .map_err(|e| LuaError::RuntimeError(e.to_string()))?
                    .available_async()
                    .await
                    .map_err(|e| LuaError::RuntimeError(e.to_string()))?
                    .into_iter()
                    .map(|v| v.version)
                    .collect::<Vec<String>>())
            })?,
        )?;
        Ok(Value::Table(table))
    }
}

impl<'lua> FromLua<'lua> for ParseLegacyFileResponse {
    fn from_lua(value: Value<'lua>, _: &'lua Lua) -> std::result::Result<Self, LuaError> {
        match value {
            Value::Table(table) => Ok(ParseLegacyFileResponse {
                version: table.get::<_, Option<String>>("version")?,
            }),
            _ => panic!("Expected table"),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::Vfox;

    #[tokio::test]
    async fn test_parse_legacy_file_nodejs() {
        let vfox = Vfox::test();
        let response = vfox
            .parse_legacy_file("vfox-nodejs", Path::new("test/data/.node-version"))
            .await
            .unwrap();
        let out = format!("{:?}", response);
        assert_snapshot!(out, @r###"ParseLegacyFileResponse { version: Some("20.0.0") }"###);
    }
}
