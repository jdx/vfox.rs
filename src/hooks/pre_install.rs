use mlua::prelude::LuaError;
use mlua::{FromLua, Lua, Value};

use crate::error::Result;
use crate::Plugin;

impl Plugin {
    #[tokio::main(flavor = "current_thread")]
    pub async fn pre_install(&self, version: &str) -> Result<PreInstall> {
        self.pre_install_async(version).await
    }

    pub async fn pre_install_async(&self, version: &str) -> Result<PreInstall> {
        let ctx = self.context(Some(version.to_string()))?;
        let pre_install = self
            .eval_async(chunk! {
                require "hooks/pre_install"
                return PLUGIN:PreInstall($ctx)
            })
            .await?;

        Ok(pre_install)
    }
}

#[derive(Debug)]
pub struct PreInstall {
    pub version: String,
    pub url: Option<String>,
    pub note: Option<String>,
    pub sha256: Option<String>,
    pub md5: Option<String>,
    pub sha1: Option<String>,
    pub sha512: Option<String>,
    // pub addition: Option<Table<'lua>>,
}

impl<'lua> FromLua<'lua> for PreInstall {
    fn from_lua(value: Value<'lua>, _: &'lua Lua) -> std::result::Result<Self, LuaError> {
        match value {
            Value::Table(table) => {
                Ok(PreInstall {
                    version: table.get::<_, String>("version")?,
                    url: table.get::<_, Option<String>>("url")?,
                    note: table.get::<_, Option<String>>("note")?,
                    sha256: table.get::<_, Option<String>>("sha256")?,
                    md5: table.get::<_, Option<String>>("md5")?,
                    sha1: table.get::<_, Option<String>>("sha1")?,
                    sha512: table.get::<_, Option<String>>("sha512")?,
                    // addition,
                })
            }
            _ => panic!("Expected table"),
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::hooks::pre_install::PreInstall;
    use crate::runtime::Runtime;
    use crate::Plugin;
    use std::string::ToString;

    #[test]
    fn dummy() {
        let pre_install = run("vfox-dummy", "1.0.1");
        assert_debug_snapshot!(pre_install);
    }

    #[test]
    fn nodejs() {
        Runtime::set_os("linux".to_string());
        Runtime::set_arch("x64".to_string());
        let pre_install = run("vfox-nodejs", "20.0.0");
        assert_debug_snapshot!(pre_install);

        Runtime::set_os("macos".to_string());
        Runtime::set_arch("arm64".to_string());
        let pre_install = run("vfox-nodejs", "20.1.0");
        assert_debug_snapshot!(pre_install);

        Runtime::set_os("windows".to_string());
        Runtime::set_arch("x64".to_string());
        let pre_install = run("vfox-nodejs", "20.3.0");
        assert_debug_snapshot!(pre_install);

        Runtime::reset();
    }

    fn run(plugin: &str, v: &str) -> PreInstall {
        let p = Plugin::test(plugin);
        p.pre_install(v).unwrap()
    }
}
