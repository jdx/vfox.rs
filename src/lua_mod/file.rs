use std::path::Path;
use std::os::unix::fs::symlink as _symlink;
use mlua::{ExternalResult, Lua, MultiValue, Table};
use crate::error::Result;

pub fn mod_file(lua: &Lua) -> Result<()> {
    let package: Table = lua.globals().get("package")?;
    let loaded: Table = package.get("loaded")?;
    Ok(loaded.set("file", lua.create_table_from(vec![
        ("symlink", lua.create_async_function(symlink)?),
    ])?)?)
}

async fn symlink<'lua>(_lua: &'lua Lua, input: MultiValue<'lua>) -> mlua::Result<()> {
    let input: Vec<String> = input.into_iter().map(|v| v.to_string()).collect::<mlua::Result<_>>()?;
    let src = Path::new(&input[0]);
    let dst = Path::new(&input[1]);
    _symlink(src, dst).into_lua_err()?;
    Ok(())
}

#[cfg(test)]
mod tests {
    use std::fs;
    use super::*;

    #[test]
    fn test_symlink() {
        let _ = fs::remove_file("/tmp/test_symlink_dst");
        let lua = Lua::new();
        mod_file(&lua).unwrap();
        lua.load(mlua::chunk! {
            local file = require("file")
            file.symlink("/tmp/test_symlink_src", "/tmp/test_symlink_dst")
        }).exec().unwrap();
        assert_eq!(fs::read_link("/tmp/test_symlink_dst").unwrap(), Path::new("/tmp/test_symlink_src"));
        fs::remove_file("/tmp/test_symlink_dst").unwrap();
    }
}