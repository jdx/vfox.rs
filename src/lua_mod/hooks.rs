use crate::error::Result;
use mlua::Lua;
use std::path::Path;

pub struct HookFunc {
    _name: &'static str,
    required: bool,
    filename: &'static str,
}

#[rustfmt::skip]
const HOOK_FUNCS: [HookFunc; 7] = [
    HookFunc { _name: "Available", required: true, filename: "available" },
    HookFunc { _name: "PreInstall", required: true, filename: "pre_install" },
    HookFunc { _name: "EnvKeys", required: true, filename: "env_keys" },
    HookFunc { _name: "PostInstall", required: false, filename: "post_install" },
    HookFunc { _name: "PreUse", required: false, filename: "pre_use" },
    HookFunc { _name: "ParseLegacyFile", required: false, filename: "parse_legacy_file" },
    HookFunc { _name: "PreUninstall", required: false, filename: "pre_uninstall" },
];

pub fn mod_hooks(lua: &Lua, root: &Path) -> Result<()> {
    for hook in &HOOK_FUNCS {
        let hook_path = root.join("hooks").join(format!("{}.lua", hook.filename));
        if hook_path.exists() {
            lua.load(hook_path).exec()?;
        } else if hook.required {
            return Err(format!("Required hook '{}' not found", hook.filename).into());
        }
    }
    Ok(())
}
