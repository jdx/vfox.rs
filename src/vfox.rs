use std::collections::BTreeMap;
use std::path::{Path, PathBuf};
use std::str::FromStr;

use itertools::Itertools;
use reqwest::Url;
use xx::file;

use crate::error::Result;
use crate::hooks::available::AvailableVersion;
use crate::hooks::env_keys::{EnvKey, EnvKeysContext};
use crate::hooks::parse_legacy_file::ParseLegacyFileResponse;
use crate::hooks::pre_install::PreInstall;
use crate::metadata::Metadata;
use crate::plugin::Plugin;
use crate::registry;
use crate::sdk_info::SdkInfo;

#[derive(Debug)]
pub struct Vfox {
    pub runtime_version: String,
    pub install_dir: PathBuf,
    pub plugin_dir: PathBuf,
    pub cache_dir: PathBuf,
    pub download_dir: PathBuf,
    pub temp_dir: PathBuf,
}

impl Vfox {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn list_available_sdks() -> &'static BTreeMap<String, Url> {
        registry::list_sdks()
    }

    pub async fn list_available_versions(&self, sdk: &str) -> Result<Vec<AvailableVersion>> {
        let sdk = self.get_sdk(sdk)?;
        sdk.available_async().await
    }

    pub fn list_installed_versions(&self, sdk: &str) -> Result<Vec<SdkInfo>> {
        let path = self.install_dir.join(sdk);
        if !path.exists() {
            return Ok(Default::default());
        }
        let versions = xx::file::ls(&path)?;
        Ok(versions
            .into_iter()
            .filter_map(|p| {
                p.file_name()
                    .and_then(|f| f.to_str())
                    .map(|s| s.to_string())
            })
            .sorted()
            .map(|version| SdkInfo::new(sdk.to_string(), version.to_string(), path.join(&version)))
            .collect())
    }
    pub fn list_sdks(&self) -> Result<Vec<Plugin>> {
        if !self.plugin_dir.exists() {
            return Ok(Default::default());
        }
        let plugins = xx::file::ls(&self.plugin_dir)?;
        plugins
            .into_iter()
            .filter_map(|p| {
                p.file_name()
                    .and_then(|f| f.to_str())
                    .map(|s| s.to_string())
            })
            .sorted()
            .map(|name| self.get_sdk(&name))
            .collect()
    }

    pub fn get_sdk(&self, name: &str) -> Result<Plugin> {
        Plugin::from_dir(&self.plugin_dir.join(name))
    }

    pub fn install_plugin(&self, sdk: &str) -> Result<Plugin> {
        let plugin_dir = self.plugin_dir.join(sdk);
        if !plugin_dir.exists() {
            let url = registry::sdk_url(sdk).ok_or_else(|| format!("Unknown SDK: {sdk}"))?;
            return self.install_plugin_from_url(url);
        }
        Plugin::from_dir(&plugin_dir)
    }

    pub fn install_plugin_from_url(&self, url: &Url) -> Result<Plugin> {
        let sdk = url
            .path_segments()
            .and_then(|s| {
                let filename = s.last().unwrap();
                filename
                    .strip_prefix("vfox-")
                    .map(|s| s.to_string())
                    .or_else(|| Some(filename.to_string()))
            })
            .ok_or("No filename in URL")?;
        let plugin_dir = self.plugin_dir.join(&sdk);
        if !plugin_dir.exists() {
            debug!("Installing plugin {sdk}");
            xx::git::clone(url.as_ref(), &plugin_dir)?;
        }
        Plugin::from_dir(&plugin_dir)
    }

    pub fn uninstall_plugin(&self, sdk: &str) -> Result<()> {
        let plugin_dir = self.plugin_dir.join(sdk);
        if plugin_dir.exists() {
            file::remove_dir_all(&plugin_dir)?;
        }
        Ok(())
    }

    pub async fn install<ID: AsRef<Path>>(
        &self,
        sdk: &str,
        version: &str,
        install_dir: ID,
    ) -> Result<()> {
        self.install_plugin(sdk)?;
        let sdk = self.get_sdk(sdk)?;
        let pre_install = sdk.pre_install(version).await?;
        let url = Url::from_str(pre_install.url.as_ref().unwrap())?;
        let file = self.download(&url, &sdk, version).await?;
        self.verify(&pre_install, &file)?;
        self.extract(&file, install_dir.as_ref())?;
        Ok(())
    }

    pub fn uninstall(&self, sdk: &str, version: &str) -> Result<()> {
        let path = self.install_dir.join(sdk).join(version);
        file::remove_dir_all(&path)?;
        Ok(())
    }

    pub async fn metadata(&self, sdk: &str) -> Result<Metadata> {
        self.get_sdk(sdk)?.get_metadata()
    }

    pub async fn env_keys(&self, sdk: &str, version: &str) -> Result<Vec<EnvKey>> {
        debug!("Getting env keys for {sdk} version {version}");
        let plugin = self.get_sdk(sdk)?;
        let path = self.install_dir.join(sdk).join(version);
        let sdk_info = SdkInfo::new(sdk.to_string(), version.to_string(), path.clone());
        let ctx = EnvKeysContext {
            args: vec![],
            version: version.to_string(),
            path,
            sdk_info: BTreeMap::from([(sdk.to_string(), sdk_info.clone())]),
            main: sdk_info,
        };
        plugin.env_keys(ctx).await
    }

    pub async fn parse_legacy_file(
        &self,
        sdk: &str,
        file: &Path,
    ) -> Result<ParseLegacyFileResponse> {
        let sdk = self.get_sdk(sdk)?;
        sdk.parse_legacy_file(file).await
    }

    async fn download(&self, url: &Url, sdk: &Plugin, version: &str) -> Result<PathBuf> {
        info!("Downloading {url}");
        let filename = url
            .path_segments()
            .and_then(|s| s.last())
            .ok_or("No filename in URL")?;
        let file = self
            .download_dir
            .join(format!("{sdk}-{version}"))
            .join(filename);
        xx::http::download(url.clone(), &file).await?;
        Ok(file)
    }

    fn verify(&self, pre_install: &PreInstall, file: &Path) -> Result<()> {
        info!("Verifying {file:?} checksum");
        if let Some(sha256) = &pre_install.sha256 {
            xx::hash::ensure_checksum_sha256(file, sha256)?;
        }
        if let Some(sha512) = &pre_install.sha512 {
            xx::hash::ensure_checksum_sha512(file, sha512)?;
        }
        if let Some(_sha1) = &pre_install.sha1 {
            unimplemented!("sha1")
        }
        if let Some(_md5) = &pre_install.md5 {
            unimplemented!("md5")
        }
        Ok(())
    }

    fn extract(&self, file: &Path, install_dir: &Path) -> Result<()> {
        info!("Extracting {file:?} to {install_dir:?}");
        let filename = file.file_name().unwrap().to_string_lossy().to_string();
        let tmp = self.temp_dir.join(&filename);
        file::remove_dir_all(&tmp)?;
        file::remove_dir_all(install_dir)?;
        let move_to_install = || {
            let subdirs = file::ls(&tmp)?;
            if subdirs.len() == 1 && subdirs.first().unwrap().is_dir() {
                let subdir = subdirs.first().unwrap();
                file::mv(subdir, install_dir)?;
                file::remove_dir_all(&tmp)?;
            } else {
                file::mv(&tmp, install_dir)?;
            }
            Result::Ok(())
        };
        if filename.ends_with(".tar.gz") {
            xx::archive::untar_gz(file, &tmp)?;
            move_to_install()?;
        } else if filename.ends_with(".tar.xz") {
            xx::archive::untar_xz(file, &tmp)?;
            move_to_install()?;
        } else if filename.ends_with(".tar.bz2") {
            xx::archive::untar_bz2(file, &tmp)?;
            move_to_install()?;
        } else if filename.ends_with(".zip") {
            xx::archive::unzip(file, &tmp)?;
            move_to_install()?;
        } else {
            unimplemented!("Unsupported extension {file:?}");
        }
        Ok(())
    }
}

impl Default for Vfox {
    fn default() -> Self {
        Self {
            runtime_version: "1.0.0".to_string(),
            plugin_dir: home().join(".version-fox/plugin"),
            cache_dir: home().join(".version-fox/cache"),
            download_dir: home().join(".version-fox/downloads"),
            install_dir: home().join(".version-fox/installs"),
            temp_dir: home().join(".version-fox/temp"),
        }
    }
}

fn home() -> PathBuf {
    homedir::my_home()
        .ok()
        .flatten()
        .unwrap_or_else(|| PathBuf::from("/"))
}

#[cfg(test)]
mod tests {
    use super::*;

    impl Vfox {
        pub fn test() -> Self {
            Self {
                runtime_version: "1.0.0".to_string(),
                plugin_dir: PathBuf::from("plugins"),
                cache_dir: PathBuf::from("test/cache"),
                download_dir: PathBuf::from("test/downloads"),
                install_dir: PathBuf::from("test/installs"),
                temp_dir: PathBuf::from("test/temp"),
            }
        }
    }

    #[tokio::test]
    async fn test_env_keys() {
        let vfox = Vfox::test();
        vfox.install_plugin("nodejs").unwrap();
        let keys = vfox.env_keys("nodejs", "20.0.0").await.unwrap();
        let output = format!("{:?}", keys).replace(
            &vfox.install_dir.to_string_lossy().to_string(),
            "<INSTALL_DIR>",
        );
        assert_snapshot!(output);
    }

    #[tokio::test]
    async fn test_install_plugin() {
        let vfox = Vfox::test();
        vfox.uninstall_plugin("nodejs").unwrap();
        assert!(!vfox.plugin_dir.join("nodejs").exists());
        vfox.install_plugin("nodejs").unwrap();
        assert!(vfox.plugin_dir.join("nodejs").exists());
    }

    #[tokio::test]
    async fn test_install() {
        let vfox = Vfox::test();
        let install_dir = vfox.install_dir.join("nodejs").join("20.0.0");
        vfox.install("nodejs", "20.0.0", &install_dir)
            .await
            .unwrap();
        assert!(vfox
            .install_dir
            .join("nodejs")
            .join("20.0.0")
            .join("bin")
            .join("node")
            .exists());
        vfox.uninstall_plugin("nodejs").unwrap();
        assert!(!vfox.plugin_dir.join("nodejs").exists());
        vfox.uninstall("nodejs", "20.0.0").unwrap();
        assert!(!vfox.install_dir.join("nodejs").join("20.0.0").exists());
        file::remove_dir_all(vfox.plugin_dir.join("nodejs")).unwrap();
        file::remove_dir_all(vfox.install_dir).unwrap();
        file::remove_dir_all(vfox.download_dir).unwrap();
    }

    #[tokio::test]
    async fn test_install_cmake() {
        let vfox = Vfox::test();
        vfox.install_plugin("cmake").unwrap();
        let install_dir = vfox.install_dir.join("cmake").join("3.21.0");
        vfox.install("cmake", "3.21.0", &install_dir).await.unwrap();
        assert!(vfox
            .install_dir
            .join("cmake")
            .join("3.21.0")
            .join("bin")
            .join("cmake")
            .exists());
        vfox.uninstall_plugin("cmake").unwrap();
        assert!(!vfox.plugin_dir.join("cmake").exists());
        vfox.uninstall("cmake", "3.21.0").unwrap();
        assert!(!vfox.install_dir.join("cmake").join("3.21.0").exists());
        file::remove_dir_all(vfox.plugin_dir.join("cmake")).unwrap();
        file::remove_dir_all(vfox.install_dir).unwrap();
        file::remove_dir_all(vfox.download_dir).unwrap();
    }

    #[tokio::test]
    async fn test_metadata() {
        let vfox = Vfox::test();
        let metadata = vfox.metadata("nodejs").await.unwrap();
        let out = format!("{:?}", metadata);
        assert_snapshot!(out);
    }
}
