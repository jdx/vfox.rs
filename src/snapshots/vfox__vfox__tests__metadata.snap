---
source: src/vfox.rs
expression: out
---
Metadata { name: "nodejs", legacy_filenames: [".node-version", ".nvmrc"], version: "0.3.0", description: Some("Node.js runtime environment."), author: None, license: Some("Apache 2.0"), homepage: Some("https://github.com/version-fox/vfox-nodejs") }
